### PR TITLE
[RFR] Suggest moves

### PR DESCRIPTION
**Pull request by Luwangel**
- [x] Implements the Taxicab algorithm with or without values
- [x] Implements the IA algorithm
- [x] Create a suggest routine to propose the best move to the user
- [x] Display the best move to the user with the H command
- [x] Fix the shuffler

### CODE REVIEW

**Review by jpetitcolas — APPROVED** [review]

**Review by jpetitcolas — CHANGES_REQUESTED** [review]
You changed shulffer, but I don't see any tests. :(

**Review by fzaninotto — COMMENTED** [review]
I'm surprised you need that, it worked well in Python without it

File: src/game/suggest.go [hunk 1]
```diff
@@ -43,6 +43,26 @@ func Taxicab(grid Grid, grid2 Grid) int {
...
+	size := len(grid)
+	y := 0
+
+	for y < size {
+		x := 0
+		for x < size {
+			if grid[y][x] != grid2[y][x] {
+				expectedPos, _ := findTileByValue(grid2, grid[y][x])
+				sum += int(math.Abs(float64(y-int(expectedPos.Y))) + math.Abs(float64(x-int(expectedPos.X))))
+				sum += size*size - int(grid[y][x])
```
Comment (jpetitcolas) [inline src/game/suggest.go:64]: Why is this line for?

File: src/game/suggest.go [hunk 2]
```diff
@@ -43,6 +43,26 @@ func Taxicab(grid Grid, grid2 Grid) int {
 	return sum
 }
 
+func TaxicabWithValues(grid Grid, grid2 Grid) int {
```
Comment (jpetitcolas) [inline src/game/suggest.go]: This method is not tested.

File: src/events/suggest_listener.go [hunk 3]
```diff
@@ -1,18 +1,31 @@
...
+	"time"
 )
 
-func SuggestListener(msgChan chan Message, gridChan chan game.Grid, startedGrid game.Grid) {
-	for {
-		grid := <-gridChan
-		path, _ := game.DeepPuzzleAlgorithm2(grid, startedGrid)
+const SUGGEST_DURATION time.Duration = 1
+
+func SuggestListener(msgChan chan Message, grid game.Grid, startedGrid game.Grid) {
```
Comment (jpetitcolas) [inline src/events/suggest_listener.go]: Where is related test? :)

**Review by jpetitcolas — CHANGES_REQUESTED** [review]

File: src/game/suggest.go (hunk 1 above)
Comment (Luwangel) [reply src/game/suggest.go:64]: It's a way to prioritize tiles with low value.

The float64 conversion is for the Math.Abs method.

File: src/game/suggest_test.go [hunk 4]
```diff
@@ -73,9 +113,334 @@ func TestCountMisplacedTiles(t *testing.T) {
...
 
-	if sum != 5 {
-		t.Error("The misplaced tiles counter of two grids should be equals to 5")
+	if sum != 35 {
+		t.Error(fmt.Sprintf("The taxicab sum with values of two different grids should be equals to 35 and not %d", sum))
+	}
+}
+
+func TestCompareTwoNodesByCostPositive(t *testing.T) {
+	grid1 := Grid{
```
Comment (jpetitcolas) [inline src/game/suggest_test.go]: Do we really need to create some filled grids as we don't use them?

**Review by jpetitcolas — CHANGES_REQUESTED** [review]

**Pull request merged**